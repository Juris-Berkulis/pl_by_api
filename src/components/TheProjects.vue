<script setup lang="ts">
import { ref, type Ref, computed, type ComputedRef, watchEffect } from 'vue';
import ProjectsList from '@/components/ProjectsList.vue';
import ProjectsListEmpty from '@/components/ProjectsListEmpty.vue'
import BaseLoader from '@/components/BaseLoader.vue';
import BasePagination from '@/components/BasePagination.vue';
import BaseSelect from '@/components/BaseSelect.vue';
import BaseInput from '@/components/BaseInput.vue';
import { type Project, type ProjectStatus, type SelectedProjectStatus } from '@/types/index';

interface FetchResult {
    count: number,
    results: Project[],
};

type ProjectsCountInEachPage = 3 | 5 | 10;

const projectsList: Ref<Array<Project>> = ref([]);
const projectsCount: Ref<number> = ref(0);
const page: Ref<number> = ref(1);
const projectsCountsInEachPage: Array<ProjectsCountInEachPage> = [3, 5, 10];
const projectsCountInEachPage: Ref<ProjectsCountInEachPage> = ref(10);
const isLoading: Ref<boolean> = ref(false);
const projectsStatusesSelectionList: Array<SelectedProjectStatus> = ['Любой', 'CREATED', 'IN_PROGRESS', 'FINISHED'];
const selectedProjectStatus: Ref<SelectedProjectStatus> = ref('Любой');
const inputedTitlePart: Ref<string> = ref('');

const pagesCount: ComputedRef<number> = computed((): number => {
    const projectsCountForLastPage: number = projectsCount.value % projectsCountInEachPage.value;
    const fulledPagesCount: number = (projectsCount.value - projectsCountForLastPage) / projectsCountInEachPage.value;
    return projectsCountForLastPage === 0 ? fulledPagesCount : fulledPagesCount + 1
});

const changePage = (newPageNumber: number): void => {
    page.value = newPageNumber;
};

const resetPage = (): void => {
    changePage(1);
};

const checkingForAllValues = (_: never): '' => {
    return ''
};

const projectsStatusForFetch: ComputedRef<ProjectStatus | ''> = computed((): ProjectStatus | '' => {
    switch (selectedProjectStatus.value) {
        case 'CREATED':
        case 'IN_PROGRESS':
        case 'FINISHED':
            return selectedProjectStatus.value
        case 'Любой':
            return ''
        default:
            return checkingForAllValues(selectedProjectStatus.value)
    }
});

const fetchProjectsList = async () => {
    isLoading.value = true;
    try {
        const response: Response = await fetch(`https://dev.aicap.tech/api/v1/interview/projects?title=${inputedTitlePart.value}&status=${projectsStatusForFetch.value}&page=${page.value}&page_size=${projectsCountInEachPage.value}`);
        if (response.ok) {
            const fetchResult: FetchResult = await response.json();
            projectsList.value = fetchResult.results;
            projectsCount.value = fetchResult.count;
        } else {
            throw 'Ошибка при запросе'
        }
    } catch (error) {
        console.log(error);
    } finally {
        isLoading.value = false;
    }
};

watchEffect(async () => {
    await fetchProjectsList();
});
</script>

<template>
<div class="flex flex-col items-center flex-1">
    <div class="w-full flex justify-between items-center mb-8 flex-wrap gap-2">
        <BaseInput v-model:inputedValue="inputedTitlePart" :resetPage="resetPage" />
        <BaseSelect v-model:selectedValue="selectedProjectStatus" :optionsList="projectsStatusesSelectionList" :resetPage="resetPage" />
        <BaseSelect v-model:selectedValue="projectsCountInEachPage" :optionsList="projectsCountsInEachPage" :resetPage="resetPage" />
    </div>
    <div class="w-full flex-1 mb-8">
        <BaseLoader v-if="isLoading" />
        <ProjectsList v-else-if="projectsList.length" :projectsList="projectsList" />
        <ProjectsListEmpty v-else />
    </div>
    <BasePagination :pagesCount="pagesCount" :changePage="changePage" :page="page" />
</div>
</template>

<style scoped>
</style>
