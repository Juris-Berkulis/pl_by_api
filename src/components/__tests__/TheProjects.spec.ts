import { mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import { ref, type Ref } from "vue";
import TheProjects from "@/components/TheProjects.vue";
import type { Project } from "@/types";

describe('TheProjects', () => {
    const wrapper = mount(TheProjects);

    it('Snapshot', () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Fetch projects list', async () => {
        const projectsList: Ref<Array<Project>> = ref([]);
        const response: Response = await fetch(`https://dev.aicap.tech/api/v1/interview/projects?page=1`);
        const fetchResult = await response.json();
        projectsList.value = fetchResult.results;

        expect(projectsList.value[0].title).toBe('Рекламный ролик №56 (банки, дебет, 2 кат)');
        expect(projectsList.value.length).toBe(10);
    });
});
