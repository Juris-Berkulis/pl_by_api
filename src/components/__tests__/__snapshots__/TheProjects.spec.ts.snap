// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TheProjects > Snapshot 1`] = `
"<div class=\\"flex flex-col items-center flex-1\\">
  <div class=\\"w-full flex justify-between items-center mb-8 flex-wrap gap-2\\"><input class=\\"w-full sm:w-1/2 px-2 py-1 border text-sm focus:outline-none\\" type=\\"text\\" placeholder=\\"Заголовок\\"><select class=\\"w-full sm:w-1/3 border p-1 text-sm focus:outline-none\\">
      <option data-option=\\"Любой\\">Любой</option>
      <option data-option=\\"CREATED\\">CREATED</option>
      <option data-option=\\"IN_PROGRESS\\">IN_PROGRESS</option>
      <option data-option=\\"FINISHED\\">FINISHED</option>
    </select><select class=\\"w-full sm:w-1/3 border p-1 text-sm focus:outline-none\\">
      <option data-option=\\"3\\">3</option>
      <option data-option=\\"5\\">5</option>
      <option data-option=\\"10\\">10</option>
    </select></div>
  <div class=\\"w-full flex-1 mb-8\\">
    <div class=\\"h-full w-full flex justify-center items-center\\">
      <div class=\\"animate-spin h-12 w-12 border-8 border-sky-200 border-r-transparent rounded-full\\"></div>
    </div>
  </div>
  <div class=\\"w-full flex justify-center gap-4\\"></div>
</div>"
`;
